### ISSUE TITLE
Remote participant cannot hear the audio of other side

### ISSUE THREAD

**Issue by papuruth** [twilio/twilio-video-app-react#285]
**Describe the bug*
We are joining the twilio room with two participant (We are allowing only two). But the remote participant cannot hear me. But when I make my self as the main seaker he can hear me and on remote side even if the remote participant doesn't make himself as the main speaker still I can hear him. I tried removing the disable audio flag from main speaker but that didn't helped.
I have modified the participant stripe and I'm showing only the participant which is not selected in the participant stripe area.(same as whatsapp video call). lso when the camera is disabled the participant cannot hear.
**To Reproduce**
1. Join the room
2. When remote participant joins I can hear him / her
3. But the remote participant cannot hear me.

**Expected behavior**
I want that no matter the speaker selected the audio should be heard whenever any participant speaks.

**Screenshots**

**Environment (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome 84.0.4147.105
 - App Version: 2.7.1
 - Node.js version: 12.16.1

**Additional context**
I don't know how but sometimes the problem doesn't arise but once I start switching the participant the issue arises.

**Comment by liorshk** [comment]
+1

**Comment by papuruth** [comment]
**Edit in description:** Now the audio issue is solved but not pefectly when I'm switching between main speaker and participant speaker the audio get detached from the DOM and doesn't re-attach but keeps replacing the same audio element when switching the user.

**Comment by timmydoza** [comment]
Thanks for the issue @papuruth!

It sounds like you are using a version of the app that you have modified for your use case. Is it possible to share some code snippets of what you have changed? Or could you share a link to a deployed version of the app? I'll be better able to help out if I could see the changes that you have made. 

Thanks!

**Comment by liorshk** [comment]
@timmydoza 
Hi, I have the same problem, can I send you a version of our app also?
Can you send me an email to lior@deep-solutions.net so I will send you the details?

**Comment by papuruth** [comment]
It would be nice if you could specify which part of code you want to see?
I cannot share the code base because it's against company policy. And the repo is hosted on gitlab on companies domain which is private.

**Comment by papuruth** [comment]
**Secondary Participant Code**
```js
import { checkEmpty } from '@/utils/commonFunctions/CommonFunctions';
import useParticipants from '@/utils/twilioVideo/hooks/useParticipants/useParticipants';
import useVideoContext from '@/utils/twilioVideo/hooks/useVideoContext/useVideoContext';
import React from 'react';
import Participant from '../Participant/Participant';
import useSelectedParticipant from '../VideoProvider/useSelectedParticipant/useSelectedParticipant';

export default function SecondaryParticipant() {
  const {
    room: { localParticipant },
  } = useVideoContext();
  const participants = useParticipants();
  const [selectedParticipant, setSelectedParticipant] = useSelectedParticipant();
  console.log('local p', localParticipant);
  console.log('selectedParticipant', selectedParticipant);
  return (
    <div className="small-video-screen">
      {!checkEmpty(participants) && (
        <Participant
          key={selectedParticipant === localParticipant ? participants[0]?.sid : localParticipant?.sid}
          videoPriority="high"
          participant={selectedParticipant === localParticipant ? participants[0] : localParticipant}
          onClick={() => setSelectedParticipant(selectedParticipant === localParticipant ? participants[0] : localParticipant)}
        />
      )}
    </div>
  );
}
```

**Main Participant Code**
```js
import useMainSpeaker from '@/utils/twilioVideo/hooks/useMainSpeaker/useMainSpeaker';
import usePublications from '@/utils/twilioVideo/hooks/usePublications/usePublications';
import useScreenShareParticipant from '@/utils/twilioVideo/hooks/useScreenShareParticipant/useScreenShareParticipant';
import React from 'react';
import styled from 'styled-components';
import MainParticipantInfo from '../MainParticipantInfo/MainParticipantInfo';
import ParticipantTracks from '../ParticipantTracks/ParticipantTracks';
import useSelectedParticipant from '../VideoProvider/useSelectedParticipant/useSelectedParticipant';

const NoVideoContainer = styled.div`
  font-size: 40px;
  color: #fff;
`;

export default function MainParticipant() {
  const mainParticipant = useMainSpeaker();
  const [selectedParticipant] = useSelectedParticipant();
  const screenShareParticipant = useScreenShareParticipant();
  const videoPriority = mainParticipant === selectedParticipant || mainParticipant === screenShareParticipant ? 'high' : null;
  const publications = usePublications(mainParticipant);
  const videoPublication = publications.find((p) => p.trackName.includes('camera'));
  const isVideoEnabled = Boolean(videoPublication);

  return (
    <MainParticipantInfo participant={mainParticipant}>
      <ParticipantTracks participant={mainParticipant} enableScreenShare videoPriority={videoPriority} />
      {!isVideoEnabled && <NoVideoContainer>Camera is off</NoVideoContainer>}
    </MainParticipantInfo>
  );
}
```

When the meeting starts the remote user is on main screen and the local user moves to secondary screen. But if I click on secondary screen the user switches and one of the audio elements gets removed from the dom and doesn't returns. After that one audio element remains in the dom and if the user is switched the audio tracks is replaced on the same element for bot user (local and remote). I have change the room seetings to allow only two user at a time in a room. One to One meeting type.

**Comment by timmydoza** [comment]
Hey @papuruth!

Thanks for sharing the code! I looked through it all and it all looks good to me. I didn't see anything in your code that would cause the issue you describe. 

Just curious, did you happen to modify the `<Participant />` or `<ParticipantTracks />` components? Problems in those components could prevent the `<AudioTrack />` component from rendering.

**Comment by arshKhachatryan** [comment]
Hi,
I have a similar issue, when I enable my audio and start screen sharing, my audio can be heard by the second user, but when I stop the screen sharing, my audio is not heard anymore. I investigated it a little and found out that when I stop my screen sharing, my audio track's medisStreamTrack's `enabled` property is set to false on the second user side(my audioTrack is remoteAudioTrack from second user perspective). In order to be heard, I am manually setting `enabled` property `participant.audioTracks.entries().next().value[1].track.mediaStreamTrack.enabled = true` to true.
@timmydoza do you have any thought why this can happen?

**Comment by rachnaban** [comment]
I am also facing the exact same issue. In my case, I have modified the code of the Participant Strip where in I am rendering only the localParticipant and the remote Participant(only one remote participant is allowed) I rendered on the main Participant div. Whenever I click on the participant from the participant strip, I swap these two participants such that their video position exchange with each other. I can only hear the audio of the remote participant when he is in the participant strip area, I mean I dont hear any audio from them when they are in the main div. 

I have removed disableAudio prop from main participant div and internally there is no code as such I see which could cause this  issue, but still I dont understand why this is happening.

@papuruth , if you are able to resolve this at your end , Can you please suggest something to me?
 Any thoughts @timmydoza  ?

**Comment by papuruth** [comment]
@rachnaban try using this code in AudioTrack.js component

```js
import useRoomState from '@/utils/twilioVideo/hooks/useRoomState/useRoomState';
import { useAppState } from '@/utils/twilioVideo/state';
import { useEffect, useRef } from 'react';

export default function AudioTrack({ track }) {
  const { activeSinkId } = useAppState();
  const roomState = useRoomState();
  const audioEl = useRef();
  useEffect(() => {
    setTimeout(() => {
      if (!document.getElementById(track?.name)) {
        audioEl.current = track.attach();
        audioEl.current.setAttribute('id', track.name);
        document.body.appendChild(audioEl.current);
      }
    }, 0);
    return () => {
      track.detach().forEach((el) => el.remove());
    };
  }, [track, roomState]);

  useEffect(() => {
    if (audioEl?.current && audioEl?.current?.setSinkId) audioEl.current.setSinkId(activeSinkId);
  }, [activeSinkId]);

  return null;
}
```

**Comment by rachnaban** [comment]
Thanks so much @papuruth for the quick response. Sure, I will try this and update here.

**Comment by rachnaban** [comment]
@papuruth , Thanks for the suggestion above. I tried , but still it is not working for me.

**Comment by papuruth** [comment]
@rachnaban could you share some code snippets of the files I have shared? So that I can see where's the problem.
Can you confirm one thing when you are switching the participants see for audio elements attached to the DOM are they getting de-attached?

**Comment by arshKhachatryan** [comment]
> Hi,
> I have a similar issue, when I enable my audio and start screen sharing, my audio can be heard by the second user, but when I stop the screen sharing, my audio is not heard anymore. I investigated it a little and found out that when I stop my screen sharing, my audio track's medisStreamTrack's `enabled` property is set to false on the second user side(my audioTrack is remoteAudioTrack from second user perspective). In order to be heard, I am manually setting `enabled` property `participant.audioTracks.entries().next().value[1].track.mediaStreamTrack.enabled = true` to true.
> @timmydoza do you have any thought why this can happen??

**Comment by timmydoza** [comment]
@rachnaban Can you look at the DOM and confirm that there is one `<audio>` element for each participant? I wonder if you are having the same issue as @papuruth where audio tracks are not being re-attached to the DOM. 

There's one thing that could be worth trying. We had a previous version of our `AudioTrack` component where the audio element is rendered by React (instead of document.appendChild). Try replacing your AudioTrack component with this:

In `src/components/AudioTrack/AudioTrack.tsx`:

```
import React, { useEffect, useRef } from 'react';
import { AudioTrack as IAudioTrack } from 'twilio-video';

interface AudioTrackProps {
  track: IAudioTrack;
}

export default function AudioTrack({ track }: AudioTrackProps) {
  const ref = useRef<HTMLAudioElement>(null!);

  useEffect(() => {
    const el = ref.current;
    track.attach(el);
    return () => {
      track.detach(el);
    };
  }, [track]);

  return <audio ref={ref} />;
}
```

**Comment by arshKhachatryan** [comment]
> Hi,
> I have a similar issue, when I enable my audio and start screen sharing, my audio can be heard by the second user, but when I stop the screen sharing, my audio is not heard anymore. I investigated it a little and found out that when I stop my screen sharing, my audio track's medisStreamTrack's `enabled` property is set to false on the second user side(my audioTrack is remoteAudioTrack from second user perspective). In order to be heard, I am manually setting `enabled` property `participant.audioTracks.entries().next().value[1].track.mediaStreamTrack.enabled = true` to true.
> @timmydoza do you have any thought why this can happen?



I troubleshot my case, and it happened that I was creating two audio elements for the same track, one was created when I enabled my audio, the second time was when I shared my screen. The problem was in this part of `ParticipantTrack` 
```
if (enableScreenShare && publications.some(p => p.trackName.includes('screen'))) {
    filteredPublications = publications.filter(p => !p.trackName.includes('camera'));
  } else {
    filteredPublications = publications.filter(p => !p.trackName.includes('screen'));
  }
```
here if screen share was enabled it entered inside the if statement, and my `filteredPublications` become every other publications except `video` publication, so here my audio and data tracks were instantiated the second time. I changed this part to this
```
if (enableScreenShare) {
    filteredPublications = publications.filter(p => p.trackName.includes('screen'));
  } else {
    filteredPublications = publications.filter(p => !p.trackName.includes('screen'));
  }
```
Hope this will help someone.

**Comment by rachnaban** [comment]
Hey @papuruth , @timmydoza and @arshKhachatryan , Thanks so much on this one. Your comments helped me a lot. This issue got fixed for me. It was my code issue. I was rendering an Avatar whenever local/remote participant was muting video, but here I forgot to attach the audio tracks back when Avatar is shown. This was the mistake! Thanks so much again.

**Comment by papuruth** [comment]
@timmydoza could you please help me in something? I want to end the call after a specified time limit automatically but I'm not where should i place my code. I tried in end call button function using react useEffect Hook and setTimeout but once the disconnects console throwing warning of can't update state on an unmounted component, where as there is no such problem if the call is ended by clicking the button. Could you please tell what could be going wrong.
Here's the code sinppet of End Call Button Function
```js
import { ReactComponent as EndCall } from '@/assets/icons/callEnd.svg';
import Toast from '@/utils/Toast';
import useVideoContext from '@/utils/twilioVideo/hooks/useVideoContext/useVideoContext';
import React, { useEffect } from 'react';
import { Fab } from 'react-tiny-fab';
import ReactToolTip from 'react-tooltip';

const styles = {
  buttonStyle: {
    color: '#000',
    backgroundColor: 'red',
  },
};

export default function EndCallButton() {
  const { room } = useVideoContext();

  const disconnectMeeting = () => {
    room.disconnect();
  };
  
  useEffect(() => {
    const t1 = setTimeout(() => {
      Toast.info('Meeting will end in 5 seconds');
    }, 5000);
    const t2 = setTimeout(() => {
      disconnectMeeting();
    }, 10000);
    return () => {
      clearTimeout(t1);
      clearTimeout(t2);
    };
  }, []);

  return (
    <>
      <Fab
        onClick={disconnectMeeting}
        mainButtonStyles={styles.buttonStyle}
        icon={(
          <p data-tip="End Call">
            <EndCall />
          </p>
        )}
        event="click"
      />
      <ReactToolTip effect="solid" type="info" offset={{ top: 10 }} backgroundColor="#72D36B" />
    </>
  );
}
```

**Comment by timmydoza** [comment]
Hey @papuruth - I'd add it in the `useRoom` hook. 

This code is from `src/components/videoContext/useRoom/useRoom.ts` starting on line 25:
(I added comments by the code I added)

```
      return Video.connect(token, { ...options, tracks: [] }).then(
        newRoom => {
          const timeoutId = setTimeout(() => newRoom.disconnect(), 10000); // <<<<<<<< Set timeouts here

          setRoom(newRoom);
          const disconnect = () => newRoom.disconnect();

          newRoom.once('disconnected', () => {
            // Reset the room only after all other `disconnected` listeners have been called.
            setTimeout(() => setRoom(new EventEmitter() as Room));
            window.removeEventListener('beforeunload', disconnect);

            clearTimeout(timeoutId); // <<<<<<<< Clear timeouts here

            if (isMobile) {
              window.removeEventListener('pagehide', disconnect);
            }
          });
```

The callback function passed to `Video.connect(...).then()` is only called once, so you don't have to worry about any components updating. Lemme know how it works out for you!

**Comment by papuruth** [comment]
@timmydoza Is this code fine? Beacause the toast is being fired multiple times and I want it to fire only once
```js
export default function useRoomState() {
  const { room } = useVideoContext();
  const [state, setState] = useState('disconnected');
  useEffect(() => {
    const setRoomState = () => setState(room.state || 'disconnected');
    setRoomState();
    room.on('disconnected', setRoomState).on('reconnected', setRoomState).on('reconnecting', setRoomState);
    room.once('participantDisconnected', (participant) => {
      Toast.info(`${participant.identity} has left the meeting.`);
    });
    room.once('participantConnected', (participant) => {
      Toast.info(`${participant.identity} has joined the meeting.`);
    });
    return () => {
      room.off('disconnected', setRoomState).off('reconnected', setRoomState).off('reconnecting', setRoomState);
      room.off('participantDisconnected', () => {}).off('participantConnected', () => {});
    };
  }, [room]);

  return state;
}
```

**Comment by timmydoza** [comment]
Sorry about the delay @papuruth. 

I think it's incorrect to include your code in the `useRoomState()` function because that function is called many times (which is why you'd see multiple toasts). 

Try putting the code in `src/components/videoContext/useRoom/useRoom.ts`. This function is only called once (after you join a room). 

```
      return Video.connect(token, { ...optionsRef.current, tracks: localTracks }).then(
        newRoom => {
          setRoom(newRoom);
          const disconnect = () => newRoom.disconnect();

          room.on('participantDisconnected', (participant) => {
            Toast.info(`${participant.identity} has left the meeting.`);
          });
          
          room.on('participantConnected', (participant) => {
            Toast.info(`${participant.identity} has joined the meeting.`);
          });
```

Let me know how this works out for you!

I'm going to go ahead and close this issue since the original issue appears to be resolved. If you have another question - feel free to open a new issue. I'm happy to help!
